docs/dev/design/file-tree-diff.rst

```diff
@@ -0,0 +1,234 @@
+
+Another alternative is to use the `difflib <https://docs.python.org/3/library/difflib.html>`__ module,
+the only downside is that it doesn't distinguish lines that were changed from lines that were added or removed.
+But maybe that's ok? Do we really need to know if a line was changed instead of added or removed?
+
+.. code:: python
+
+   import difflib
+
+   diff = difflib.ndiff(["one", "two", "three", "four"], ["ore", "three", "four", "five"])
+   print(list(diff))
+   # ['+ ore', '- one', '- two', '  three', '  four', '+ five']
+
+A good thing of using Python is that we don't need to write the files to disk,
+and the result is easier to parse.
+
+Storing results
+---------------
+
+Doing a diff between two versions can be expensive, so we need to store the results.
+
+We can store the results in the DB (``VersionDiff``), or in S3 (``diff/project/version-a/version-b.json``), or maybe a combination of both.
+The information to store would contain some information about the versions compared, the build, and the diff itself.
+
+.. code:: json
+
+   {
+       "version_a": {"id": 1, "build": {"id": 1}},
+       "version_b": {"id": 2, "build": {"id": 2}},
+       "diff": {
+           "added": [{"file": "new.txt"}],
+           "removed": [{"file": "deleted.txt"}],
+           "modified": [{"file": "changed.txt", "lines": {"added": 1, "removed": 1}}]
+       }
+   }
+
+The information is stored in a similar way that it will be returned by the API.
+Things important to note:
+
+- We need to take into consideration the diff of the latest successful builds only.
+  If any of the builds from the stored diff don't match the latest successful build of any of the versions,
+  we need to the diff again.
+- Once we have the diff between versions ``A`` and ``B``, we can infer the diff between ``B`` and ``A``.
+  We can store that information as well, or just calculate it on the fly.
+- The list of files are objects, so we can store additional information in the future.
+- When a file has been modified, we also store the number of lines that changed.
+  We could also show this for files that were added or removed.
+- Instead of using the slugs on storage (if we decide to use S3), we could use the IDs, that will prevent the need to update the data if the slugs change.
+  But if the slugs change, the builds will probably be different, so we will need to update the data anyway.
+- If a project or version is deleted (or deactivated), we should delete the diff as well.
+
+We could store the changed files sorted by the number of changes, or make that an option in the API,
+or just let the client sort the files as they see fit.
+
+A combination of using the DB and S3 would be really useful,
+as the model will serve as the lock for the API,
+and will be easier to keep track of which build was used to generate the diff
+without having to download the diff file from S3.
+
+API
+---
+
+This operation is expensive, so it won't be available to unauthenticated users.
```

Comment:
The action will have an RTD token that it uses to query the information, or use the webhook approach.

Reply:
So, every user will have to set this token when setting up the GH Action?

Reply:
I think asking the user to setup a Read the Docs token for this, is too much. Besides, we don't even support "tokens with limited capabilities", so they will need to use their personal token for this, which is a no go to me.

I would avoid requiring authentication for this endpoint. We need to expose the files changed to all users, even un-authed ones -- in particular for the public projects with addons integration.

Reply:
I agree making this authed seems like a bad place to start. I'd much rather make the endpoint be less expensive. I don't really understand why this would be slow if we already have VersionDiff objects in the DB? It seems like it should just be a DB query.

Reply:
> I don't really understand why this would be slow if we already have VersionDiff objects in the DB? It seems like it should just be a DB query.

The first time the diff is done will be slow, someone could try to abuse this to generate a diff for all possible versions.

Reply:
Also, to use this API for private projects, we do need auth for the GH action or whatever that hits the API.